### PR TITLE
[air] add tuner user guide

### PR DESCRIPTION
**Pull request by xwjiang2010**
## Why are these changes needed?

User guide for Tuner on the Ray AIR docs.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/ray-air/tuner.rst [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+.. literalinclude:: doc_code/tuner.py
+    :language: python
+    :start-after: __tuner_start__
+    :end-before: __tuner_end__
+
+.. note::
+    ``num_samples = 2`` here will be applied to the whole suite of grid search. In other words,
+    we will generate 4 trials.
+
+For a more end-to-end example, checkout "analyze_tuning_results" example.
```
Comment (xwjiang2010) [inline doc/source/ray-air/tuner.rst:73]: Should we demote this to just an example rather than user guide?
Also how do I link into it?

Comment (richardliaw) [reply doc/source/ray-air/tuner.rst:73]: link via `:ref:`

File: doc/source/ray-air/tuner.rst [hunk 2]
```diff
@@ -0,0 +1,91 @@
+.. _air-tuner:
+
+Hyperparameter Tuning
+=====================
+Tuner API is the recommended way of launching hyperparameter tuning jobs with Ray Tune.
```
Comment (richardliaw) [inline doc/source/ray-air/tuner.rst:5]: Add "what is a tuner and what does it do"?

File: doc/source/ray-air/user-guides.rst [hunk 3]
```diff
@@ -22,6 +22,18 @@ AIR Feature Guides
...
         :classes: btn-link btn-block stretched-link
+
+
+    ---
+    :img-top: /ray-overview/images/ray_svg_logo.svg
+
+    +++
+    .. link-button:: /ray-air/tuner
+        :type: ref
+        :text: How to use AIR Tuner?
```
Comment (ericl) [inline doc/source/ray-air/user-guides.rst:33]:
```suggestion
        :text: Hyperparameter Tuning
```

We are now keeping them consistent with titles in master.

File: doc/source/_toc.yml [hunk 4]
```diff
@@ -15,6 +15,7 @@ parts:
     - file: ray-air/user-guides
       sections:
         - file: ray-air/preprocessors
+        - file: ray-air/tuner
```
Comment (ericl) [inline doc/source/_toc.yml:18]: Can you order to right before "How to Analyze Tune Results" in master?

**Review by krfricke — APPROVED** [review]
LGTM let's follow up with the restore docs soon.

Also cc @richardliaw for review

File: doc/source/ray-air/tuner.rst [hunk 5]
```diff
@@ -0,0 +1,140 @@
+.. _air-tuner:
+
+Hyperparameter Tuning
+=====================
+The Ray AIR Tuner API is the recommended way to launch hyperparameter tuning jobs with Ray Tune.
```
Comment (richardliaw) [inline doc/source/ray-air/tuner.rst:5]: I think we should phrase this differently. In here it sounds like Ray AIR Tuner is a wrapper over Ray Tune, but actually they should be one and the same. 

How about saying "The Tuner API is the recommended of tuning hyperparameters in Ray AIR."?

File: doc/source/ray-air/tuner.rst [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+In this guide, we will show you how to do this.
+
+.. note::
+   The Tuner API can also be used to launch hyperparameter tuning without using Ray AIR Trainers - see
+   :ref:`the Ray Tune documentation <tune-main>` for more guides examples.
+
+Basic usage
+-----------
+
+Assuming you already have your own Trainer, this is how the Ray Tune API works in a nutshell:
```
Comment (richardliaw) [inline doc/source/ray-air/tuner.rst:24]:
```suggestion
Suppose that you already have a Ray AIR trainer at hand. You have fitted it and obtained a model and training results.

Below, we demonstrate how you can plug in your existing Trainer into a Tuner.
```

File: doc/source/ray-air/tuner.rst [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+.. literalinclude:: doc_code/tuner.py
+    :language: python
+    :start-after: __basic_start__
+    :end-before: __basic_end__
+
+
+Defining the search space
+-------------------------
+
+The Tuner API takes in a `param_space` argument where you can define the search space that Ray Tune will
```
Comment (richardliaw) [inline doc/source/ray-air/tuner.rst:35]: I'm thinking we should not explicitly reference Ray Tune as a separate entity

File: doc/source/ray-air/tuner.rst [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+use to try out hyperparameter configurations.
```
Comment (richardliaw) [inline doc/source/ray-air/tuner.rst:36]:
```suggestion
`Tuner` takes in a `param_space` argument where you can define the search space from which hyperparameter configurations will be sampled.
```

File: doc/source/ray-air/tuner.rst [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+    .. literalinclude:: doc_code/tuner.py
+        :language: python
+        :start-after: __torch_start__
+        :end-before: __torch_end__
+
+
+As you can see in the above example, the Tuner API allows you to specify a search space over most parameters:
+
+- Parameters will be passed to the Trainer
```
Comment (richardliaw) [inline doc/source/ray-air/tuner.rst:71]: A couple gotchas about the behavior of parameters:

File: doc/source/ray-air/tuner.rst [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+In general, all the arguments accepted by your :ref:`Trainer <air-trainer-ref>` can be tuned
+(except for the :class:`ray.air.config.RunConfig`).
+
+
+Specify TuneConfig
+------------------
+This config contains tuning specific settings, including the tuning algorithm to use, the metric and mode to rank results etc.
+
+See the :class:`Tuner <ray.tune.tuner.Tuner>` and the :class:`TuneConfig API reference <ray.tune.tune_config.TuneConfig>` for more details.
```
Comment (richardliaw) [inline doc/source/ray-air/tuner.rst:98]: Add a basic example here?

File: doc/source/ray-air/tuner.rst [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+
+Specify RunConfig
+-----------------
+This config contains framework's runtime configurations that are more generic than tuning specific settings.
+This may include failure/retry configurations, verbosity levels, the name of the experiment, its logging directory,
+checkpoint configuration as well as its syncing configuration.
+
+See the :class:`RunConfig API reference <ray.air.config.RunConfig>` for more details.
```
Comment (richardliaw) [inline doc/source/ray-air/tuner.rst:107]: Add a basic example here?

File: doc/source/ray-air/tuner.rst [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+.. literalinclude:: doc_code/tuner.py
+    :language: python
+    :start-after: __tuner_start__
+    :end-before: __tuner_end__
+
+.. note::
+    ``num_samples=2`` here will be applied to the whole suite of grid search. In other words,
+    we will generate 4 trials.
+
+For a more end-to-end example, checkout our :doc:`/ray-air/examples/analyze_tuning_results` guide.
```
Comment (richardliaw) [inline doc/source/ray-air/tuner.rst:122]: This guide only for analysis right?

File: doc/source/ray-air/tuner.rst [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+The Tuner API takes in a `param_space` argument where you can define the search space that Ray Tune will
+use to try out hyperparameter configurations.
+
+Generally, you can search over most arguments and configurations provided by Ray AIR (and your own options).
+This includes Ray Datasets, Preprocessors, the distributed training configuration (ScalingConfig)
+and general hyperparameters.
+
+The main exclusion here are all arguments of the :class:`ray.air.config.RunConfig`, which are inherently un-tunable.
+Another exclusion is the :class:`ray.tune.tune_config.TuneConfig`, which defines the tuning behavior itself - the settings
+apply to all trials and thus can't be tuned.
```
Comment (richardliaw) [inline doc/source/ray-air/tuner.rst:44]: Let's move this to the bottom of this subsection.

**Review by richardliaw — CHANGES_REQUESTED** [review]
looking good; just a couple comments.

File: doc/source/ray-air/tuner.rst (hunk 12 above)
Comment (richardliaw) [reply doc/source/ray-air/tuner.rst:122]: We probably want to link out to Ray Tune user guide instead?

**Review by richardliaw — APPROVED** [review]

**Comment by richardliaw** [comment]
I just did an editing pass here and added a graphic. It looks good to me; hope to merge by tomorrow.

File: doc/source/ray-air/tuner.rst [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+
+Configuring Hyperparameter Tuning
+=================================
+
+The Ray AIR :class:`Tuner <ray.tune.Tuner>` is the recommended way to tune hyperparameters in Ray AIR.
+
+.. image:: images/tuner.svg
+    :align: center
+
+`Tuner` provides an interface that works with AIR Trainers to perform distributed
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:11]: A `Tuner`

File: doc/source/ray-air/tuner.rst [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+hyperparameter tuning. `Tuner` is part of Ray Tune and provides
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:12]: As part of Ray Tune, the `Tuner` provides ...

File: doc/source/ray-air/tuner.rst [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+a variety of state-of-the-art hyperparameter tuning algorithms for optimizing model
+performance.
+
+This page will cover basic usage examples of a Tuner. If you are interested in
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:16]: Not a fan of "This page will..."

I would rephrase as: 

"What follows next is basic coverage of what a Tuner is and how you can use it for basic examples.

File: doc/source/ray-air/tuner.rst [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+reading more, please take a look at the :ref:`Ray Tune documentation <tune-main>`.
+
+Key Concepts
+------------
+
+There are a couple key concepts for using a Tuner:
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:22]: A think there is a list of items here not a couple. If you going to enumerate key concepts perhaps start with this:

There are a number of key concepts that dictate proper use of a Tuner:

File: doc/source/ray-air/tuner.rst [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+
+* First, you need to specify the hyperparameters you want to tune in a `search space`.
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:24]: Remove first. Not needed.

 * A specific set of hyperparameters in a `search space.`
 * A `search algorithm` to ...
 * A set of arguments (`search space`, `search algorithm`, `scheduler`, and `Trainer`) passed to the `Tuner`
 *  Each individual hyperparater evalution run is called a `trial`
 * The `Tuner` returns its results in a `ResultGrid` object.

File: doc/source/ray-air/tuner.rst [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+----------------------------------
+
+A `Tuner` takes in a `param_space` argument where you can define the search space
+from which hyperparameter configurations will be sampled.
+
+Depending on the model and dataset, you may want to tune:
+
+- The training batch size
+- The learning rate for SGD-based training (e.g. image classification)
+- The maximum depth for tree-based models (e.g. XGBoost)
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:57]: e.g.,

File: doc/source/ray-air/tuner.rst [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+not limited to:
+
+- Ray Datasets
+- Preprocessors
+- Scaling configurations
+- and other hyperparameters.
+
+There are a couple gotchas about parameter specification when using Tuners with Trainers:
+
+- By default, configuration dictionaries and config objects will be deep-merged
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:87]: deep-merged.

File: doc/source/ray-air/tuner.rst [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+- Parameters that are duplicated in the Trainer and Tuner will be overwritten by the Tuner ``param_space``
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:88]: ``param-space``.

File: doc/source/ray-air/tuner.rst [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+- **Exception:** all arguments of the :class:`RunConfig <ray.air.config.RunConfig>` and :class:`TuneConfig <ray.tune.tune_config.TuneConfig>` are inherently un-tunable.
+
+
+How to configure a Tuner
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:92]: How to configure a Tuner?

File: doc/source/ray-air/tuner.rst [hunk 23]
```diff
@@ -0,0 +1,216 @@
...
+
+- failure/retry configurations
+- verbosity levels
+- the name of the experiment
+- the logging directory
+- checkpoint configurations
+- custom callbacks
+- integration with cloud storage
+
+The following we showcase some common configuration of :class:`RunConfig <ray.air.config.RunConfig>`.
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:123]: In the following section, we cover and showcase some common ...

File: doc/source/ray-air/tuner.rst [hunk 24]
```diff
@@ -0,0 +1,216 @@
...
+
+Below, we demonstrate how you can use a Trainer object with a Tuner.
+
+.. literalinclude:: doc_code/tuner.py
+    :language: python
+    :start-after: __basic_start__
+    :end-before: __basic_end__
+
+
+How do I configure a search space?
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:47]: How to configure a search space?

I would leave out first person I here.

File: doc/source/ray-air/tuner.rst [hunk 25]
```diff
@@ -0,0 +1,216 @@
...
+
+.. literalinclude:: doc_code/tuner.py
+    :language: python
+    :start-after: __run_config_start__
+    :end-before: __run_config_end__
+
+See the :class:`RunConfig API reference <ray.air.config.RunConfig>` for more details.
+
+
+How do I specify parallelism?
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:133]: How to configure parallelism?

File: doc/source/ray-air/tuner.rst [hunk 26]
```diff
@@ -0,0 +1,216 @@
...
+
+.. literalinclude:: doc_code/tuner.py
+    :language: python
+    :start-after: __tune_parallelism_start__
+    :end-before: __tune_parallelism_end__
+
+Read more about this in :ref:`tune-parallelism` section.
+
+
+How do I specify an optimization algorithm?
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:153]: How to specify an optimization algorithm?

File: doc/source/ray-air/tuner.rst [hunk 27]
```diff
@@ -0,0 +1,216 @@
...
+    :language: python
+    :start-after: __tune_optimization_start__
+    :end-before: __tune_optimization_end__
+
+Read more about this in the :ref:`Search Algorithm <search-alg-ref>` and :ref:`Scheduler <schedulers-ref>` section.
+
+How to analyze results?
+-----------------------
+
+``Tuner.fit()`` generates a ResultGrid object. This object contains metrics, results, and checkpoints
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:171]: `ResultGrid`

File: doc/source/ray-air/tuner.rst [hunk 28]
```diff
@@ -0,0 +1,216 @@
...
+of each trial. Below is a simple usage example:
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:172]: Below is a simple example:

File: doc/source/ray-air/tuner.rst [hunk 29]
```diff
@@ -0,0 +1,216 @@
...
+
+.. literalinclude:: doc_code/tuner.py
+    :language: python
+    :start-after: __result_grid_inspection_start__
+    :end-before: __result_grid_inspection_end__
+
+Advanced Tuning
+---------------
+
+Tuners also offer the possibility to apply different data preprocessing steps, as shown in the following snippet.
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:182]: should this be the "capability" instead of "possibility"?

File: doc/source/ray-air/tuner.rst [hunk 30]
```diff
@@ -0,0 +1,216 @@
...
+
+.. literalinclude:: doc_code/tuner.py
+    :language: python
+    :start-after: __tune_preprocess_start__
+    :end-before: __tune_preprocess_end__
+
+You can also sample different train/validation datasets:
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:189]: Additionally, you sample different ...

File: doc/source/ray-air/tuner.rst [hunk 31]
```diff
@@ -0,0 +1,216 @@
...
+
+
+Restoring and resuming
+----------------------
+A Tuner regularly saves its state, so that a tuning run can be resumed after being interrupted.
+
+Additionally, if trials fail during a tuning run, they can be retried - either from scratch or
+from the latest available checkpoint.
+
+To restore the Tuner state, you just pass the path to the experiment directory to ``Tuner.restore()``.
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:204]: To restore the Tuner state, pass the path to the experiment directory as an argument to ``Tuner.restore(...)``

File: doc/source/ray-air/tuner.rst [hunk 32]
```diff
@@ -0,0 +1,216 @@
...
+
+This path can be obtained e.g. from the output of a tuning run (it's called "Result logdir"). If you
```
Comment (dmatrix) [inline doc/source/ray-air/tuner.rst:206]: This path is obtained from the output of a tuning run, namely "Result logdir." However, if you specify a ``name`` in the :class:`RunConfig <ray.air.config.RunConfig>`, then it is located under ``~/ray_results/<name>``.

**Review by dmatrix — COMMENTED** [review]
I'm bit confused about the diagram. Why it has Two boxes named "Trainer." We may want to clarify that in the caption.

**Comment by richardliaw** [comment]
@dmatrix addressed comments!

**Review by dmatrix — APPROVED** [review]
